File: python/p0008_string_to_integer.py
class Solution:
    def myAtoi(self, str: str) -> int:
        str = str.lstrip()
        if len(str) == 0:
            return 0

        sign = ''
        if str[0] == '-' or str[0] == '+':
            sign = str[0]
            str = str[1:]

        res = 0
        for i in range(len(str)):
            if not str[i].isdigit():
                break
            res = res * 10 + int(str[i])

        if sign == '-':
            res = -res
            
        if res < -2 ** 31:
            return -2 ** 31
        elif res > 2 ** 31 - 1:
            return 2 ** 31 - 1
        else:
            return res

File: src/solutions/p0008_string_to_integer.rs
struct Solution;

impl Solution {
    pub fn my_atoi(str: String) -> i32 {
        let mut number = vec![];

        for (i, c) in str.trim().chars().enumerate() {
            if !c.is_digit(10) {
                if i != 0 || c != '-' {
                    break;
                }
                if c == '+' {
                    continue;
                }
            }
            number.push(c.to_string())
        }

        if let Ok(n) = number.join("").parse::<i32>() {
            return n;
        }

        if number.len() <= 1 {
            0
        } else if number[0] == "-" {
            std::i32::MIN
        } else {
            std::i32::MAX
        }
    }
}
